Use typed coordinates more

paulrouget opened the pull request:
Requires https://github.com/servo/servo/pull/19895

We use Size2D and Point2D across compositing, constellation and script, loosing the type of pixels we use (DevicePixel, DeviceIndepententPixel or CSSPixel) along the way, which might lead to bugs like `window.outerHeight` not taking into account the page zoom (using DeviceIndepententPixel instead of CSSPixel).

This should make the situation a bit better.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [x] These changes do not require tests because we can't zoom in a test

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20071)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/webdriver_server/lib.rs, components/script/dom/screen.rs, components/constellation/constellation.rs, components/script/dom/window.rs
 * @cbrewster: components/constellation/constellation.rs
 * @jgraham: components/webdriver_server/lib.rs
 * @fitzgen: components/script/dom/screen.rs, components/script_traits/script_msg.rs, components/script/dom/window.rs, components/script_traits/lib.rs
 * @kichjang: components/script/dom/screen.rs, components/script_traits/script_msg.rs, components/script/dom/window.rs, components/script_traits/lib.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

paulrouget commented on components/constellation/constellation.rs:
```diff
@@ -2551,14 +2551,16 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
+        if let Some(top_level_browsing_context_id) = top_level_browsing_context_id {
+            let browsing_context_id = BrowsingContextId::from(top_level_browsing_context_id);
+            self.resize_browsing_context(new_size, size_type, browsing_context_id);
+        }
```
That was necessary because the initial size of the window would not be sent to the constellation, and the page would assume, at load, a pixel ratio of 1.0 (default when the size is not set yet).

paulrouget commented:
Not ready for review. The screen coordinates are not correct.

paulrouget commented:
I moved from `u32` to `usize` because it's easier to cast a `TypedSize<f32>` to `usize` (via `TypedSize::to_usize`) than `u32` (`let s: TypedSize<u32> = TypedSize::cast().unwrap()`).

paulrouget commented:
Back to WIP. Need https://github.com/servo/euclid/pull/274 and #19895

paulrouget committed:
Use typed coordinates.

We use Size2D and Point2D across compositing, constellation and script,
losing the type of pixels we use (DevicePixel, DeviceIndepententPixel
or CSSPixel) along the way, which might lead to bugs like
window.outerHeight not taking into account the page zoom (using
DeviceIndepententPixel instead of CSSPixel).

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20153) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
@glennw r?

glennw commented:
We seem to pass around the generic types a lot, which makes things a bit harder to read.

How about we define a set of type aliases for each of the unit / type combinations, similar to how units.rs works in WR? (https://github.com/servo/webrender/blob/master/webrender_api/src/units.rs#L24)

paulrouget commented:
@glennw Sure. I'm not against that. But I like when types are explicit :)

I'll add a commit that aliases these types.

paulrouget commented:
@glennw so I tried to use these aliased types. We use Typed dimensions a lot all across Servo. I feel like we should either use these aliases everywhere, or nowhere. Also, that means adding some new few types (CSSSize/Lenght/Rect, DeviceIndependentSize/Length/Rect) all over.

So, to be consistent, I have to modify many files across all the components.

If you think it's important, I'll make the changes all over Servo, but I'm not sure it's necessary.

What do you think?

glennw commented:
We wouldn't need to use them everywhere immediately though, right? Couldn't we just use the aliased types in this commit and then start using them more incrementally in other parts of the code? (Since the aliases are not considered a separate type by rustc, there's no cast / conversion necessary between an alias and a TypedSize of the same units).

paulrouget committed:
alias some euclid types

glennw commented:
Reviewed 2 of 13 files at r3, 11 of 11 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/compositing/windowing.rs, line 126 at r4](https://reviewable.io:443/reviews/servo/servo/20071#-L7XBu9D3i7WZNYf3x_N:-L7XBu9D3i7WZNYf3x_O:b81o60v) ([raw file](https://github.com/servo/servo/blob/26739a67a3ebe4c1e76c3d73a8bb2cf739a7fa9d/components/compositing/windowing.rs#L126)):*
> ```Rust
>     fn window_rect(&self) -> DeviceUintRect;
>     /// Returns the size of the window.
>     fn size(&self) -> DeviceSize;
> ```

Was the comment here wrong or have the semantics changed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/20071)*
<!-- Sent from Reviewable.io -->

glennw commented:
Looks ok to me apart from the one question above. It'd be good to manually check hit testing for at least a device pixel ratio of 1 and 2, since it's easy to miss mistakes here.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20303) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget committed:
remove WindowMethods::size()

paulrouget commented:
> Was the comment here wrong or have the semantics changed?

The comment was right, but we changed the type, from DeviceIndependentPixel to DevicePixel. It's actually not used anywhere except in the sibling function `framebuffer_size`. So let's remove that method.

> It'd be good to manually check hit testing for at least a device pixel ratio of 1 and 2

Tested on my laptop and and on phone with OS' ratio. Also tested by change the ratio via the OS preference, and also via `--device-pixel-ratio`.

glennw commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit fef0506 has been approved by `glennw`

<!-- @bors-servo r=glennw fef0506269495545c5ee9ca5957904bbcb130bb9 -->

bors-servo commented:
:hourglass: Testing commit fef0506269495545c5ee9ca5957904bbcb130bb9 with merge fc90e613d8851b663e3304d713b2ca08aa397bf4...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11058), [arm32](http://build.servo.org/builders/arm32/builds/11295), [arm64](http://build.servo.org/builders/arm64/builds/11328), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11366), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8247), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8262), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11422), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3816), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3844), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8090), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8157), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3917), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3864), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6757)
Approved by: glennw
Pushing fc90e613d8851b663e3304d713b2ca08aa397bf4 to master...
